File: .agents/mcp-servers/antigravity-docs/mcp_server.js
import { Server } from "@modelcontextprotocol/sdk/server/index.js";
import { StdioServerTransport } from "@modelcontextprotocol/sdk/server/stdio.js";
import {
  CallToolRequestSchema,
  ListToolsRequestSchema,
  ListResourcesRequestSchema,
  ReadResourceRequestSchema,
} from "@modelcontextprotocol/sdk/types.js";
import fs from "fs";
import path from "path";
import { fileURLToPath } from "url";

// Emulate __dirname
const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);
const KB_DIR = path.join(__dirname, "knowledge_base");

// Create MCP server instance
const server = new Server(
  {
    name: "Antigravity_Docs",
    version: "1.0.0",
  },
  {
    capabilities: {
      resources: {},
      tools: {},
    },
  }
);

/**
 * Handle listing available documentation pages as Resources
 */
server.setRequestHandler(ListResourcesRequestSchema, async () => {
  if (!fs.existsSync(KB_DIR)) {
    return { resources: [] };
  }

  const files = fs.readdirSync(KB_DIR).filter((file) => file.endsWith(".md"));
  const resources = files.map((file) => ({
    uri: `docs://index/${file}`,
    name: `Antigravity Docs: ${file}`,
    mimeType: "text/markdown",
    description: `Full documentation content for ${file}`,
  }));

  return { resources };
});

/**
 * Handle reading a specific documentation page
 */
server.setRequestHandler(ReadResourceRequestSchema, async (request) => {
  const uri = request.params.uri;
  const match = uri.match(/^docs:\/\/index\/(.+)$/);

  if (!match) {
    throw new Error(`Invalid URI format. Expected docs://index/<filename>`);
  }

  const filename = match[1];
  const filePath = path.join(KB_DIR, filename);

  if (!fs.existsSync(filePath)) {
    throw new Error(`File ${filename} not found.`);
  }

  const content = fs.readFileSync(filePath, "utf-8");

  return {
    contents: [
      {
        uri,
        mimeType: "text/markdown",
        text: content,
      },
    ],
  };
});

/**
 * Handle listing tools (Search keyword)
 */
server.setRequestHandler(ListToolsRequestSchema, async () => {
  return {
    tools: [
      {
        name: "search_docs",
        description:
          "Searches the Antigravity documentation for a specific keyword or phrase.",
        inputSchema: {
          type: "object",
          properties: {
            query: {
              type: "string",
              description: "The keyword or phrase to search for.",
            },
          },
          required: ["query"],
        },
      },
      {
        name: "list_document_names",
        description:
          "Returns the exact filenames of all available Antigravity documentation pages located in the knowledge base.",
        inputSchema: {
          type: "object",
          properties: {},
        },
      },
    ],
  };
});

/**
 * Handle execution of tools
 */
server.setRequestHandler(CallToolRequestSchema, async (request) => {
  if (request.params.name === "list_document_names") {
    if (!fs.existsSync(KB_DIR)) {
      return {
        content: [{ type: "text", text: "Knowledge base directory not found." }],
      };
    }
    const files = fs.readdirSync(KB_DIR).filter((file) => file.endsWith(".md"));
    return {
      content: [
        {
          type: "text",
          text: `Available documents:\n${files.join("\n")}`,
        },
      ],
    };
  }

  if (request.params.name === "search_docs") {
    const query = String(request.params.arguments?.query || "").toLowerCase();

    if (!fs.existsSync(KB_DIR)) {
      return {
        content: [{ type: "text", text: "Knowledge base directory not found." }],
      };
    }

    const files = fs.readdirSync(KB_DIR).filter((file) => file.endsWith(".md"));
    const matches = [];

    for (const filename of files) {
      const filePath = path.join(KB_DIR, filename);
      const content = fs.readFileSync(filePath, "utf-8").toLowerCase();
      if (content.includes(query)) {
        matches.push(filename);
      }
    }

    if (matches.length === 0) {
      return {
        content: [{ type: "text", text: `No results found for '${query}'.` }],
      };
    }

    let resultText = `Found '${query}' in the following documents:\n`;
    for (const match of matches) {
      resultText += `- ${match}\n`;
    }
    resultText += `\nUse read_resource(docs://index/<filename>) to read the full content.`;

    return {
      content: [{ type: "text", text: resultText }],
    };
  }

  throw new Error("Tool not found");
});

// Run server using standard I/O transport
async function startServer() {
  const transport = new StdioServerTransport();
  await server.connect(transport);
  console.error("Antigravity Docs MCP Server running on stdio");
}

startServer().catch((error) => {
  console.error("Fatal error:", error);
  process.exit(1);
});

File: .agents/mcp-servers/antigravity-docs/mcp_server.py
import os
import json
from mcp.server.fastmcp import FastMCP

# Create the MCP server instance
mcp = FastMCP("Antigravity_Docs")

KB_DIR = "knowledge_base"

@mcp.resource("docs://index")
def list_docs() -> str:
    """Returns an index of all available Antigravity documentation pages."""
    if not os.path.exists(KB_DIR):
        return "Knowledge base directory not found."
    
    files = sorted(os.listdir(KB_DIR))
    index = "# Antigravity Documentation Index\n\n"
    for f in files:
        if f.endswith(".md"):
            index += f"- {f}\n"
    return index

@mcp.tool()
def search_docs(query: str) -> str:
    """Searches the Antigravity documentation for a specific keyword or phrase."""
    if not os.path.exists(KB_DIR):
        return "Knowledge base directory not found."
        
    results = []
    for filename in sorted(os.listdir(KB_DIR)):
        if filename.endswith(".md"):
            path = os.path.join(KB_DIR, filename)
            with open(path, "r", encoding="utf-8") as f:
                content = f.read()
                if query.lower() in content.lower():
                    results.append(filename)
                    
    if not results:
        return f"No results found for '{query}'."
        
    response = f"Found '{query}' in the following documents:\n"
    for r in results:
        response += f"- {r}\n"
    response += "\nUse read_doc(filename) to read the full content of a specific page."
    return response

@mcp.tool()
def read_doc(filename: str) -> str:
    """Reads the full content of a specific documentation file. 
    Provide the exact filename, e.g., '01_Home.md'."""
    path = os.path.join(KB_DIR, filename)
    if not os.path.exists(path):
        return f"Error: File '{filename}' not found."
        
    with open(path, "r", encoding="utf-8") as f:
        return f.read()

if __name__ == "__main__":
    # This runs the MCP server via standard input/output (stdio)
    print("Starting Antigravity Docs MCP Server...")
    mcp.run()
